crates/telio-nurse/src/aggregator.rs

```diff
@@ -0,0 +1,689 @@
+use std::{collections::HashSet, time::Duration};
+
+use tokio::time::Instant;
+
+use parking_lot::RwLock;
+use serde::{ser::SerializeTuple, Serialize, Serializer};
+
+use telio_crypto::PublicKey;
+use telio_model::{
+    api_config::{EndpointProvider, FeatureNurse},
+    HashMap,
+};
+use telio_utils::telio_log_warn;
+use telio_wg::{
+    uapi::{AnalyticsEvent, PeerState},
+    WireGuard,
+};
+
+// Possible endpoint connection types
+// Their combinations give us strategies number to 15
+#[derive(Copy, Clone, Debug, Eq, Hash, PartialEq)]
+pub enum EndpointType {
+    #[allow(dead_code)]
+    Relay = 0,
+    Local = 1,
+    Stun = 2,
+    UPnP = 3,
+}
+
+impl From<EndpointProvider> for EndpointType {
+    fn from(value: EndpointProvider) -> Self {
+        match value {
+            EndpointProvider::Local => EndpointType::Local,
+            EndpointProvider::Stun => EndpointType::Stun,
+            EndpointProvider::Upnp => EndpointType::UPnP,
+        }
+    }
+}
+
+#[derive(Copy, Clone, Debug, Eq, Hash, PartialEq, Serialize)]
+pub enum RelayConnectionState {
+    Connecting = 16,
+    Connected = 17,
+}
+
+impl From<PeerState> for RelayConnectionState {
+    fn from(value: PeerState) -> Self {
+        if let PeerState::Connected = value {
+            RelayConnectionState::Connected
+        } else {
+            if let PeerState::Disconnected = value {
+                telio_log_warn!("Got disconnected state for Relay server");
+            }
+            RelayConnectionState::Connecting
+        }
+    }
+}
+
+#[derive(Copy, Clone, Debug, Eq, Hash, PartialEq, Serialize)]
+pub struct RelayConnectionData {
+    state: RelayConnectionState,
+    reason: RelayConnectionChangeReason,
+}
+
+// Possible disconnection reasons
+#[derive(Copy, Clone, Debug, Eq, Hash, PartialEq, Serialize)]
+#[allow(dead_code)]
+pub enum RelayConnectionChangeReason {
+    // 1XX reason numbers for configuration changes
+    ConfigurationChange = 101,
+    DisabledByUser = 102,
+
+    // 2XX reason numbers for network problems
```

Comment:
I wonder, maybe it is worth incorporating some sort of OS error (e.g. errno) when `NetworkError` occurs? :thinking:

Reply:
Ok, lets leave them for future improvements if we will indeed need them.